### PR TITLE
feat: eszip v2 wasm library

### PR DESCRIPTION
**Pull request by littledivy**
## API

```typescript
export interface Parser {
  parse(reader: ReadableStreamByobReader): Promise<string[]>;
  parseBytes(bytes: Uint8Array): Promise<string[]>;
  
  load(): Promise<void>;

  getModuleSource(specifier: string): Promise<string>;
  getModuleSourceMap(specifier: string): Promise<string>;
};

export function build(roots: string[], loader: () => LoadResult): Uint8Array;
```

### a BYOB readable stream

```typescript
import { Parser } from "./lib/eszip_wasm.generated.js";

const res = await fetch("http://localhost:8000/redirect.eszip2");
const reader = res.body!.getReader({ mode: "byob" });
const parser = new Parser();
const specifiers = await parser.parse(reader);

await parser.load();
for (const specifier of specifiers) {
   await parser.getModuleSource(specifier);
}
```

### Uint8Array

```typescript
import { Parser } from "./lib/eszip_wasm.generated.js";

const res = await Deno.readFile("src/testdata/redirect.eszip2");
const parser = new Parser();
const specifiers = await parser.parseBytes(res);

await parser.load();
for (const specifier of specifiers) {
   await parser.getModuleSource(specifier);
}
```

## Building eszips

```typescript
const eszip = await build([
  "https://example.com/mod.ts",
  "https://example.com/a.ts",
], (specifier: string) => {
  if (specifier === "https://example.com/a.ts") {
    return {
      specifier,
      headers: {
        "content-type": "text/typescript",
      },
      content: "export const a = 1;",
    }
  };

  return {
    specifier: "https://example.com/mod.ts",
    headers: {
      "content-type": "application/typescript",
    },
    content: `import "https://example.com/a.ts";`,
  };
});

assert(eszip instanceof Uint8Array);
```

### CODE REVIEW

File: lib/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+license = "MIT"
+
+[lib]
+name = "eszip_wasm"
+path = "lib.rs"
+crate-type = ["cdylib"]
+
+[dependencies]
+js-sys = { version = "0.3.55" }
+web-sys = { version = "0.3.56", features = ["ReadableStreamByobReader", "console"] }
```
Comment (littledivy) [inline lib/Cargo.toml:16]: Self note: remove `console`

**Comment by lucacasonato** [comment]
@littledivy We actually don't need the parser in WASM. We just need the builder (deno_graph -> eszip serialized buffer)

**Comment by littledivy** [comment]
Too late :stuck_out_tongue:, It's done. Let's keep it if you don't mind? I'll add the graph -> eszip encoder too

**Comment by lucacasonato** [comment]
> Too late 😛, It's done. Let's keep it if you don't mind? I'll add the graph -> eszip encoder too

Sure, let's keep it if it already is done :-)

File: lib/lib.rs [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+          | deno_graph::MediaType::TypeScript
+          | deno_graph::MediaType::Mts
+          | deno_graph::MediaType::Tsx
+          | deno_graph::MediaType::Dts
+          | deno_graph::MediaType::Dmts => {
+            let parsed_source = module.maybe_parsed_source().as_ref().unwrap();
+            let TranspiledSource {
+              text: source,
+              source_map: maybe_source_map,
+            } = parsed_source.transpile(Default::default())?;
```
Comment (littledivy) [inline lib/lib.rs:305]: Needs https://github.com/denoland/deno_graph/pull/124

Comment (littledivy) [reply lib/lib.rs]: Resolved

**Review by lucacasonato — COMMENTED** [review]
Add

```
[workspace]
members = [
  "lib",
]
```

to the root Cargo.toml please, and remove lib/Cargo.lock

**Review by littledivy — COMMENTED** [review]
Do not merge. WASM needs to be compiled with `--release`

File: Cargo.toml [hunk 3]
```diff
@@ -7,6 +7,15 @@ repository = "https://github.com/denoland/eszip"
 description = "A utility that can download JavaScript and TypeScript module graphs and store them locally in a special zip file"
 license = "MIT"
 
+[workspace]
+members = ["lib"]
+
+[profile.release]
+codegen-units = 1
+incremental = true
```
Comment (lucacasonato) [inline Cargo.toml:15]: Why incremental release?

**Review by lucacasonato — APPROVED** [review]
Let's land it!

**Pull request merged**